File: libcpp/src/md-gfm-doc.cc
#include <logia/md-gfm-doc.hh>

namespace logia {

MdGfmDoc::MdGfmDoc(const std::string &name, bool link_inr_root)
    : Document(name, repr_document_type(), link_inr_root) {}

void MdGfmDoc::code_begin(const std::string &language) {
  raw_os().flush();
  raw_os() << "```" << language << "\n";
}

void MdGfmDoc::code_end() {
  raw_os() << "```\n";
  raw_os().flush();
}

MdGfmDoc::OsWrapperHelper<MdGfmDoc::CodeHelper>
MdGfmDoc::code(const std::string &language) {
  code_begin(language);
  return OsWrapperHelper<CodeHelper>(*this);
}

void MdGfmDoc::image(const std::string &label, const std::string path) {
  raw_os() << "![" << label << "](" << path << ")\n";
  raw_os().flush();
}

void MdGfmDoc::doc_link(const std::string &doc_id, const std::string &text) {
  raw_os() << "[" << text << "](#!doc-" << doc_id << ")";
  raw_os().flush();
}

void MdGfmDoc::doc_link(const Document &doc, const std::string &text) {
  doc_link(doc.id(), text.empty() ? doc.name() : text);
  raw_os().flush();
}

} // namespace logia

File: libpython/logia/mdgfmdoc.py
from .document import Document

TYPE = "cmark-gfm"

class MdGfmDoc(Document):

    def __init__(self, name, link_in_root):
        super().__init__(name, TYPE, link_in_root)

    def header(self, level, title):
        self.write('{} {}\n'.format('#' * level, title))
        self.flush()
        
    def h1(self, title): self.header(1, title)
    def h2(self, title): self.header(2, title)
    def h3(self, title): self.header(3, title)
    def h4(self, title): self.header(4, title)

    def code_begin(self, lang = ""):
        self.flush()
        self.write("```{}\n".format(lang))

    def code_end(self):
        self.write("```\n")
        self.flush()

    # Insert an image (path may be local or global url)
    def img(self, label, path):
        self.write("![{}]({})\n".format(label, path))

    # Add a link to another document
    # doc_id may be the str id, or directly the document object
    def doc_link(self, doc_id, text = ""):
        if not isinstance(doc_id, str):
            if text == "":
                text = doc_id.conf.name
            doc_id = doc_id.conf.id

        self.write("[{}](#!doc-{})".format(text, doc_id))
        

    
